src/algebra/algebraic_card.lean

```diff
@@ -0,0 +1,105 @@
+-/
+
+import analysis.complex.basic
+import data.complex.cardinality
+import data.polynomial.cardinal
+import ring_theory.algebraic
+import topology.bases
+
+/-!
+### Cardinality of algebraic numbers
+
+In this file, we prove the following result: the cardinality of algebraic numbers under an R-algebra
+that's also a T1 topological space with basis s is at most `# polynomial R * # s`. From this, we
+deduce that real and complex algebraic numbers have cardinality `ω`.
+
+Although this can be used to prove transcendental numbers exist, a more direct proof is given by
+`liouville.is_transcendental`.
+-/
+
+open cardinal polynomial topological_space
+open_locale cardinal
+
+theorem algebraic_card (R) {A} [comm_ring R] [is_domain R] [comm_ring A] [is_domain A] [algebra R A]
+  [no_zero_smul_divisors R A] [topological_space A] [t1_space A] {s : set (set A)}
+  (hs : is_topological_basis s) : #{x : A | is_algebraic R x} ≤ #(polynomial R) * #s :=
+begin
+  classical,
+  apply @mk_le_of_surjective (polynomial R × s) {x : A | is_algebraic R x} (λ ⟨p, t, ht⟩,
+    if hr : p ≠ 0 ∧ ∃ x : A, x ∈ t ∩ p.root_set A
+    then ⟨classical.some hr.2, p, hr.1, (mem_root_set_iff hr.1 _).1 (classical.some_spec hr.2).2⟩
+    else ⟨0, is_algebraic_zero⟩),
+  rintro ⟨x, p, hp, he⟩,
+  suffices : ∃ t ∈ s, t ∩ p.root_set A = {x},
+  { rcases this with ⟨t, hts, ht⟩,
+    have hx := set.mem_singleton x,
+    have H : ¬p = 0 ∧ ∃ x, x ∈ t ∩ p.root_set A := ⟨hp, x, by rwa ←ht at hx⟩,
+    use [p, t, hts],
+    simp_rw dif_pos H,
+    simpa [ht, set.mem_singleton_iff] using classical.some_spec H.2 },
+  { have H : is_open (p.root_set A \ {x})ᶜ := begin
+      rw is_open_compl_iff,
+      refine set.finite.is_closed (set.finite.inter_of_left _ _),
+      exact p.root_set_finite A
+    end,
+    rw [compl_sdiff, hs.is_open_iff] at H,
+    rcases H x (set.mem_union_right _ (set.mem_singleton x)) with ⟨t, ht, hts, hxt⟩,
+    use [t, ht],
+    rw set.eq_singleton_iff_unique_mem,
+    use [hts, (mem_root_set_iff hp _).2 he],
+    rintros y ⟨hyt, hy⟩,
+    cases hxt hyt with hy' hy',
+    { exact (hy' hy).elim },
+    { exact hy' } }
+end
+
+theorem algebraic_card' (R) {A} [comm_ring R] [is_domain R] [comm_ring A] [is_domain A]
+  [algebra R A] [no_zero_smul_divisors R A] [topological_space A] [t1_space A] {s : set (set A)}
+  (hs : is_topological_basis s) : #{x : A | is_algebraic R x} ≤ max (#R) ω * #s :=
+(algebraic_card R hs).trans (mul_le_mul_right' polynomial.cardinal_mk_le_max _)
+
+theorem algebraic_card_of_second_countable (R A : Type*) [comm_ring R] [is_domain R] [comm_ring A]
+  [is_domain A] [algebra R A] [no_zero_smul_divisors R A] [topological_space A] [t1_space A]
+  [second_countable_topology A] : #{x : A | is_algebraic R x} ≤ max (#R) ω :=
+begin
+  rcases exists_countable_basis A with ⟨s, hs', _, hs⟩,
+  apply ((algebraic_card' R hs).trans ((mul_le_mul_left' ((mk_set_le_omega s).2 hs') _))).trans,
+  rw mul_omega_eq (le_max_right _ _)
+end
```

Comment:
Here's an alternate proof:
```
lemma key_lemma {α β : Type*} (c : cardinal) (f : α → β) (hf : ∀ b : β, #(f ⁻¹' {b}) ≤ c) :
  #α ≤ #β * c :=
begin
  revert hf,
  refine cardinal.induction_on c (λ γ hf, _),
  let g : (Σ b : β, f ⁻¹' {b}) → β × γ := λ ⟨b, a⟩, ⟨b, (hf b).some a⟩,
  have hg : function.injective g := by
  { rintros ⟨b₁, a₁⟩ ⟨b₂, a₂⟩ h,
    have : b₁ = b₂ := (prod.ext_iff.mp h).1,
    subst this,
    exact sigma.subtype_ext rfl (subtype.ext_iff.mp ((hf b₁).some.inj' (prod.ext_iff.mp h).2)) },
  calc #α = #Σ b : β, f⁻¹' {b} : mk_congr (equiv.sigma_preimage_equiv f).symm
  ... ≤ #β * #γ : mk_le_of_injective hg,
end

lemma algebraic.cardinal_mk_le_mul
  {R A : Type*} [comm_ring R] [is_domain R] [comm_ring A] [is_domain A] [algebra R A]
  [no_zero_smul_divisors R A] : #{x : A | is_algebraic R x} ≤ #(polynomial R) * ω :=
begin
  classical,
  let S := {x : A | is_algebraic R x},
  let g : S → polynomial R := λ x, classical.some x.2,
  refine key_lemma ω g (λ f, _),
  suffices : fintype (g ⁻¹' {f}),
  { exact @mk_le_omega _ (@fintype.encodable _ this) },
  by_cases hf : f = 0,
  { rw hf,
    have key : g ⁻¹' {0} = ∅,
    { exact set.eq_empty_iff_forall_not_mem.mpr (λ x, (classical.some_spec x.2).1), },
    rw key,
    exact set.fintype_empty },
  let h : g ⁻¹' {f} → f.root_set A := λ x, ⟨x.1.1, begin
    refine (mem_root_set_iff hf x.1.1).mpr _,
    have key' : g x  = f := x.2,
    simp_rw ← key',
    exact (classical.some_spec x.1.2).2,
  end⟩,
  exact fintype.of_injective h (λ _ _, subtype.ext ∘ subtype.ext ∘ subtype.ext_iff.mp),
end

lemma algebraic.cardinal_mk_le_max
  {R A : Type*} [comm_ring R] [is_domain R] [comm_ring A] [is_domain A] [algebra R A]
  [no_zero_smul_divisors R A] : #{x : A | is_algebraic R x} ≤ max (#R) ω :=
calc #{x : A | is_algebraic R x} ≤ #(polynomial R) * ω : algebraic.cardinal_mk_le_mul
... ≤ max (max (#(polynomial R)) ω) ω : mul_le_max (#(polynomial R)) ω
... = max (#(polynomial R)) ω : max_eq_left (le_max_right (#(polynomial R)) ω)
... ≤ max (#R) ω : max_le cardinal_mk_le_max (le_max_right (#R) ω)
```

Reply:
I've PRed the cardinal lemma: #13147

Reply:
Thanks! This is a way better proof than what I had in mind. I'll adapt the PR with this proof and your other suggestions.